### PR TITLE
Completed Undo/Redo Functionality

### PR DESCRIPTION
**Pull request by elenaconn**
### **Types of Changes**

- [x] Added functionality (snapShotFunc, line 35) in Canvas.tsx to take snapshots of State to store in the Past array (Interfaces.ts), then added snapShotFunc into the useDrop function (line 49, Canvas.tsx)
- [x] Added Undo case in componentReducer.ts (line 624) for Undo functionality in the canvas of ReacType application, which undoes HTML tag drops onto the canvas
- [x] Added Redo case in componentReducer.ts (line 649) for Redo functionality in the canvas of ReacType application, which redoes HTML tag drops onto the canvas
- [x] Added snapShotFunc (line 21) and the invocation of snapShotFunc (line 58) in DirectChildHTMLNestable.tsx to capture a snapshot of State every time an HTML tag is dropped onto the canvas to store in the past array for Undo and Redo functionality
- [x] Added Undo and Redo buttons in RightContainer.tsx (lines 522-543)
- [x] Added undoAction and redoAction actions in RightContainer.tsx (lines 204 & 208) for invocation upon clicking the Undo and Redo buttons in the right panel
- [x] Added past and future properties to the State interface in Interfaces.ts (lines 16 & 17)
- [x] Added past and future properties to the initialState object in initialState.ts (lines 25 & 26) to store snapshots of state for Undo/Redo functionality
- [x] Added form tags to the removeSeparators function in TreeChart.tsx to remove separators from forms to prevent them from rendering in the tree chart (HTML tag nesting)

### **Purpose**

- Created functionality for Undo and Redo buttons for draggable HTML tags on the canvas
- Stored snapshots of State in past and future arrays to grab recently dragged HTML tags on the canvas for undoing/redoing

### **Screenshots**

**snapShotFunc takes snapshots of state to be stored in the past and future arrays for undo/redo functionality in Canvas.tsx**
Screen Shot 2021-03-09 at 10 55 02 PM

**Undo case in componentReducer.ts**
Screen Shot 2021-03-09 at 10 56 18 PM

**Redo case in componentReducer.ts**
Screen Shot 2021-03-09 at 10 58 02 PM

**Undo/Redo button rendering in RightContainer.tsx**
Screen Shot 2021-03-09 at 11 00 05 PM

**Undo/Redo properties in State interface in Interfaces.ts**
Screen Shot 2021-03-09 at 11 01 06 PM

### CODE REVIEW

**Comment by eddypjr** [comment]
Love the detailed notes, great work frontend team!

**Comment by slhoehn** [comment]
looks good!

**Review by slhoehn — APPROVED** [review]

**Review by eddypjr — APPROVED** [review]
Looks good, all thats left is to clean up console.logs and commented code!  

Great work frontend team.

File: app/src/helperFunctions/customHook.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  useEffect(() => {
+    ref.current = value;
+    // console.log('useEffect', ref.current);
+    // return ref.current;
+  });
+  // console.log('ref.current', ref.current)
+  return ref.current;
+}
+
+export default usePrevious;
```
Comment (khuongdn16) [inline app/src/helperFunctions/customHook.ts:16]: Discussed with Edward and Elena and confirmed this file will be removed with another commit.

**Review by khuongdn16 — CHANGES_REQUESTED** [review]
Looks good. Please check my comment.

File: app/src/helperFunctions/customHook.ts (hunk 1 above)
Comment (elenaconn) [reply app/src/helperFunctions/customHook.ts:16]: On it!

**Review by khuongdn16 — APPROVED** [review]
Looks good!
